WordPress/gutenberg
# [RNMobile] Fix dictation regression on iOS

Proposed fix for https://github.com/wordpress-mobile/gutenberg-mobile/issues/5165.

Related PRs:

* `Gutenberg Mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5612
* `WordPress iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/20669

## What?
The changes in this PR address an issue with dictation not working as expected on devices running iOS 16 or later.

## Why?
Following [improvements made to dictation](https://www.macrumors.com/guide/ios-16-siri/) in iOS 16‌, [a previous hack](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5165#issuecomment-1410257774) that avoided an empty string being propagated with dictation, stopped working.

Now, with devices that are running iOS 16 or later, text added with iOS dictation is lost when tapping into the editor or typing while the dictation is still recording. This causes unexpected content loss, and could be viewed as an accessibility issue for those who rely on dictation to use the editor.

## How?
By removing the previous dictation-related hacks, we fallback to the default dictation system, which works without causing any content loss for later iOS versions. The reasoning behind following this approach is detailed [here](https://github.com/WordPress/gutenberg/pull/49452#issuecomment-1568725546).

In addition, it's still necessary to remove the `obj` symbol that's added due to [an RN bug](https://github.com/facebook/react-native/issues/36521). This is done [directly within the textViewDidChange() function](https://github.com/WordPress/gutenberg/blob/7cd7a0f53efc0af629ac211e10bd689fcc124e23/packages/react-native-aztec/ios/RNTAztecView/RCTAztecView.swift#L768-L771). 

## Testing Instructions
An installable build is available at https://github.com/wordpress-mobile/WordPress-iOS/pull/20669 for testing on a physical device.

The following steps need to be tested with both iOS 16 or later as well as on an earlier version of iOS:


<details>
 <summary>Test 1: Ensure new blocks don't replace old ones ⤵️</summary>

1. Navigate to My Site → Posts and create a new post.
2. Type some text into a paragraph block.
3. Tap the inserter in the editor's toolbar.
4. Select a new block to add to the editor.
5. Confirm that the new block is added beneath the currently select block, and that it doesn't replace it.

Based on testing steps from this PR: https://github.com/WordPress/gutenberg/pull/49154
</details>

<details>
 <summary>Test 2: Verify there is no content loss while dictating ⤵️</summary>

1. Navigate to My Site → Posts and create a new post.
2. Activate the iOS dictation feature and dictate some text.
3. Notice the text added to the editor.
4. If the keyboard remains visible while dictating, tap another block on the editor or begin typing onto the keyboard. Alternatively, if the keyboard is not visible while dictating, tap the language pop up.
5. Verify that taking an action while dictation is still running doesn't lead to content loss

Based on bug testing steps from this issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/5165
</details>

<details>
 <summary>Test 3: Verify no obj symbol is added after dictation ⤵️</summary>

1. Navigate to My Site → Posts and create a new post.
2. Tap on the Post Title.
3. Start dictating with voice.
4. When finished, press Enter.
5. Confirm that there is no obj symbol appended to the post title.

Based on testing steps from this PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1969
</details>

<details>
 <summary>Test 4: Verify block expands with content ⤵️</summary>

1. Navigate to My Site → Posts and create a new post.
2. Tap on "Start writing..." to focus the default paragraph block.
3. Say "I am writing a new block with my voice"
4. Switch to keyboard input and type some text
5. Confirm that the paragraph box expands to a second line

Only happens with an external keyboard (only applicable to iPad or Simulator with physical keyboard input)
</details>

<details>
 <summary>Test 5: Verify text position is retained after dictating in middle of paragraph ⤵️</summary>

1. Navigate to My Site → Posts and create a new post.
2. Tap on "Start writing..." to focus the default paragraph block.
3. Activate the iOS dictation feature and dictate "Mary had a little lamb its fleece was white as snow"
4. Place the cursor after "little" then deactivate iOS dictation
5. Activate the iOS dictation feature and dictate "Old MacDonald had a Farm"
7. Tap below the block to start a new paragraph
8. Confirm that the ordering of the words is "Mary had a little old MacDonald had a Farm lamb it's fleece was white as snow" _not_ "Mary had a little lamb it's fleece was white as snow old MacDonald had a Farm"

Based on this PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1969
</details>

In addition, verify that there are no failed tests across the Gutenberg, Gutenberg Mobile, and WordPress iOS PRs signalling unwanted side effects.

## Screenshots or screencast 

The following screencast shows dictation working, even after tapping the language pop up and resuming dictation (previously this led to content loss):

https://github.com/WordPress/gutenberg/assets/2998162/61c4967b-4c3b-4c83-9dab-3075af7d53fc
